generative/maisi/scripts/diff_model_infer.py

```diff
@@ -0,0 +1,222 @@
+
+    local_rank = int(os.environ["LOCAL_RANK"])
+    world_size = int(dist.get_world_size())
+    device = torch.device("cuda", local_rank)
+    torch.cuda.set_device(device)
+    print(f"Using {device} of {world_size}")
+
+    # Set random seed for reproducibility
+    if args.diffusion_unet_inference["random_seed"] == None:
+        random_seed = random.randint(0, 99999)
+    else:
+        random_seed = args.diffusion_unet_inference["random_seed"] + local_rank
+    set_determinism(random_seed)
+    print(f"random seed: {random_seed}")
+
+    output_size = tuple(args.diffusion_unet_inference["dim"])
+    out_spacing = tuple(args.diffusion_unet_inference["spacing"])
+
+    ckpt_filepath = f"{args.model_dir}/{args.model_filename}"
+    output_prefix = args.output_prefix
+
+    if local_rank == 0:
+        print(f"[config] ckpt_filepath -> {ckpt_filepath}.")
+        print(f"[config] random_seed -> {random_seed}.")
+        print(f"[config] output_prefix -> {output_prefix}.")
+        print(f"[config] output_size -> {output_size}.")
+        print(f"[config] out_spacing -> {out_spacing}.")
+
+    # Initialize autoencoder
+    autoencoder = define_instance(args, "autoencoder_def").to(device)
+    try:
+        checkpoint_autoencoder = load_autoencoder_ckpt(args.trained_autoencoder_path)
+        autoencoder.load_state_dict(checkpoint_autoencoder)
+    except:
+        print("The trained_autoencoder_path does not exist!")
+
+    # Initialize UNet model
+    unet = define_instance(args, "diffusion_unet_def").to(device)
+    checkpoint = torch.load(f"{ckpt_filepath}", map_location=device)
+    unet.load_state_dict(checkpoint["unet_state_dict"], strict=True)
+    print(f"checkpoints {ckpt_filepath} loaded.")
+
+    num_downsample_level = 1
+    if isinstance(args.diffusion_unet_def["num_channels"], list):
+        num_downsample_level = max(num_downsample_level, len(args.diffusion_unet_def["num_channels"]))
+    elif isinstance(args.diffusion_unet_def["attention_levels"], list):
+        num_downsample_level = max(num_downsample_level, len(args.diffusion_unet_def["attention_levels"]))
+    print(f"num_downsample_level -> {num_downsample_level}.")
+
+    divisor = 2 ** (num_downsample_level - 2)
+    print(f"divisor -> {divisor}.")
+
+    num_train_timesteps = checkpoint["num_train_timesteps"]
+    print(f"num_train_timesteps -> {num_train_timesteps}.")
+
+    noise_scheduler = define_instance(args, "noise_scheduler")
+    noise_scheduler.set_timesteps(num_inference_steps=args.diffusion_unet_inference["num_inference_steps"])
+
+    scale_factor = checkpoint["scale_factor"]
+    print(f"scale_factor -> {scale_factor}.")
+
+    autoencoder.eval()
+    unet.eval()
+
+    recon_model = ReconModel(autoencoder=autoencoder, scale_factor=scale_factor).to(device)
+
+    with torch.no_grad():
+        with torch.cuda.amp.autocast(enabled=True):
+            noise = torch.randn(
```

Comment:
The use of `torch.randn((t,)).to(device)` could be simplified to `torch.randn([t], device=device)`.

Reply:
Hi @mingxin-zheng, I have updated the script. Please proceed. Thanks.